File: EstructurasLineales/C++/Deque.cpp
#include "Deque.h"

template <typename T>
Deque<T>::Deque() {
	front = NULL;
	rear = NULL;
	size = 0;
}

template <typename T>
Deque<T>::Deque(T d) {
	DLNode<T> *node = new DLNode<T>(d);
	front = rear = node;
	size = 1;
}

template <typename T>
Deque<T>::~Deque() {}

template <typename T>
void Deque<T>::enqueueFront(T d) {
	DLNode<T> *node = new DLNode<T>(d);
	if (isEmpty()) {
		front = rear = node;
	}else{
		node->next = front;
		front->back = node;
		front = node;
	}
	size++;
}

template <typename T>
void Deque<T>::enqueueRear(T d) {
	if (isEmpty()) {
		enqueueFront(d);
	}
	else {
		DLNode<T> *node = new DLNode<T>(d);
		rear->next = node;
		node->back = rear;
		rear = node;
		size++;
	}
}

template <typename T>
T Deque<T>::dequeueFront() {
	if (!isEmpty()) {
		DLNode<T> *fd = front;//Declarando un apuntador y asignandole el valor que tiene data en el apuntador front
		if (size == 1) {
			clearDeque();
		} else {
			front = front->next;
			front->back = NULL;
			fd->next = NULL;//NAMAS PORQUE SI
			size--;
		}
		return fd->data;
	} else {
		return NULL;
	}
}

template <typename T>
T Deque<T>::dequeueRear() {
	if (!isEmpty()) {
		DLNode<T> *rd = rear;
		if (size == 1) {
			clearDeque();
		}
		else {
			rear = rear->back;
			rear->next = NULL;
			rd->back = NULL;//ESTA GENTE
			size--;
		}
		return rd->data;
	} else {
		return NULL;
	}
}

template <typename T>
T Deque<T>::getFront() {
		return front->data;
}

template <typename T>
T Deque<T>::getRear() {
		return rear->data;
}

template <typename T>
int Deque<T>::getSize() {
	return size;
}

template <typename T>
bool Deque<T>::isEmpty() {
	return size == 0;
}

template <typename T>
void Deque<T>::clearDeque() {
	front = rear = NULL;
	size = 0;
}

template <typename T>
void Deque<T>::showDeque() {
	std::cout << "front->";
	DLNode<T>* p = front;
	if (!p) std::cout << "NULL";
	while (p) {
		p->print();
		p = p->next;
	}
	std::cout << "<-rear" << endl;
}

File: EstructurasLineales/Java/Deque.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package ed.list;

/**
 *
 * @author A01411440
 * @param <T>
 */
public class Deque <T>{
    private NodeDL<T> front;
    private NodeDL<T> rear;
    private int size;
    
    Deque(){
        front = null;
	rear = null;
	size = 0;
    }
    Deque(T d){
        NodeDL node = new NodeDL(d);
        front = rear = node;
        size = 1;
    }
    
    public void enqueueFront(T d){
        NodeDL node = new NodeDL(d);
	if (isEmpty()) {
		front = rear = node;
	}else{
		node.next = front;
		front.prev = node;
		front = node;
	}
	size++;
    }
    
    public void enqueueRear(T d){
        if (isEmpty()) {
		enqueueFront(d);
	}
	else {
		NodeDL node = new NodeDL(d);
		rear.next = node;
		node.prev = rear;
		rear = node;
		size++;
	}
    }
    
    public T dequeueFront(){
        if (!isEmpty()) {
		NodeDL<T> fd = front;//Declarando un apuntador y asignandole el valor que tiene data en el apuntador front
		if (size == 1) {
			clearDeque();
		} else {
			front = front.next;
			front.prev = null;
			fd.next = null;//NAMAS PORQUE SI
			size--;
		}
		return fd.data;
	} else {
		return null;
	}
    }
    
    public T dequeueRear(){
        if (!isEmpty()) {
		NodeDL<T> rd = rear;
		if (size == 1) {
			clearDeque();
		}
		else {
			rear = rear.prev;
			rear.next = null;
			rd.prev = null;//ESTA GENTE
			size--;
		}
		return rd.data;
	} else {
		return null;
	}
    }
    
    public T getFront(){
        return front.data;
    }
    
    public T getRear(){
        return rear.data;
    }
    
    public int getSize(){
        return size;
    }
    
    public boolean isEmpty(){
        return size == 0;
    }
    
    public void clearDeque(){
        front = rear = null;
        size = 0;
    }
    
    public void showDeque(){
        System.out.print("front->"); 
        NodeDL<T> p =front;
        if(p==null){
            System.out.print("NULL");
        }
        while(p!=null){
            p.print();
            p=p.next;
        }
        System.out.println("<-rear");
    }
}
